### PR TITLE
 Introduce `ConnectorMetadata#getColumns`

### PR DESCRIPTION
**Pull request by ssheikin**
Use `getColumns` instead of `getColumnHandles` when ordered list of
`ColumnHandle` is needed.

Despite all known implementations of
`Map<String, ColumnHandle> getColumnHandles` return `ImmutableMap`,
`Map` interface does not provide ordering guarantees. Therefore
it is safer to rely on `List<ColumnHandle> getColumns`.

found during https://github.com/trinodb/trino/pull/6745#discussion_r566142132

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -897,13 +897,13 @@ public ConnectorInsertTableHandle beginRefreshMaterializedView(ConnectorSession
...
-            table.getSchemaName(),
-            table.getTableName(),
-            SchemaParser.toJson(icebergTable.schema()),
-            PartitionSpecParser.toJson(icebergTable.spec()),
-            getColumns(icebergTable.schema(), typeManager),
-            getDataPath(icebergTable.location()),
-            getFileFormat(icebergTable));
+                table.getSchemaName(),
+                table.getTableName(),
+                SchemaParser.toJson(icebergTable.schema()),
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:912]: let's move this to a separate PR and merge quickly

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 2]
```diff
@@ -195,6 +195,16 @@ default ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connec
         return emptyList();
     }
 
+    /**
+     * Gets all of the columns on the specified table, or an empty list if the columns cannot be enumerated.
+     *
+     * @throws RuntimeException if table handle is no longer valid
+     */
+    default List<ColumnHandle> getColumns(ConnectorSession session, ConnectorTableHandle tableHandle)
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:203]: Per commit message, this is complementary to `Map<String, ColumnHandle> getColumnHandles` -- it reduces less information, but provides an explicit ordering.

Do we want to maintain the two methods for ever?
Or can we consider deprecating one of them and removing over time?
The design proposed here does not allow this (none of the variants is strictly superior over the other), but maybe there are other options we could consider?

**Comment by findepi** [comment]
cc @martint @losipiuk

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java (hunk 2 above)
Comment (losipiuk) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:203]: I have mixed feelings about this one. It is nice that we are explicit about data structure semantics by using `List` where we want expose an ordered collection. 

On the other hand I see also drawbacks. Major one is that we now do not have a single source of truth. 
If connectors sometimes return inconsistent information via `getColumns` and `getColumnHandles` it will result hard to track failures.
Also with `getColumnHandles` in place, being used to old patters, we can still write code which iterates over map, which would no longer be correct.

Minor thing is that naming does not reflect the semantics, but I assume we could work out better names.

Comment (ssheikin) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:203]: > Do we want to maintain the two methods for ever?

May be yes, but I have an idea to introduce another method 
```
Map<ColumnHandle, ColumnMetadata> getColumnMetadatas
``` 
which probably can be used further to emulate 
```
Map<String, ColumnHandle> getColumnHandles
```
with path `ColumnHandle` -> `ColumnMetadata` -> `column_name` -> Map<column_name, ColumnHandle>
but it is to investigate if it brings performance issues.

Comment (ssheikin) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:203]: Agree.

> we can still write code which iterates over map, which would no longer be correct.

It will be 'correct' to exactly the same level as it was before with 'unforced guarantee' that 'Map' is backed by 'ImmutableMap'.

Part of this discussion was here: https://github.com/trinodb/trino/pull/6745#discussion_r566084730

> Minor thing is that naming does not reflect the semantics, but I assume we could work out better names.

Happy to apply your suggestions.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 1 above)
Comment (ssheikin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:912]: https://github.com/trinodb/trino/pull/6912

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 3]
```diff
@@ -195,6 +195,16 @@ default ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connec
         return emptyList();
     }
 
+    /**
+     * Gets all of the columns on the specified table, or an empty list if the columns cannot be enumerated.
```
Comment (kokosing) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:199]: I don't get this `or`. Is it designed for metadata queries or analyzer? If the latter we fail in case of error.

File: core/trino-main/src/main/java/io/trino/connector/system/SystemTablesMetadata.java [hunk 4]
```diff
@@ -101,10 +103,25 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     }
 
     @Override
-    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    public List<ColumnHandle> getColumns(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        return ImmutableList.copyOf(getTableColumns(session, tableHandle));
+    }
+
+    private ImmutableList<SystemColumnHandle> getTableColumns(ConnectorSession session, ConnectorTableHandle tableHandle)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/connector/system/SystemTablesMetadata.java:111]: please move it below the usages.

**Review by kokosing — COMMENTED** [review]
> Despite all known implementations of
Map<String, ColumnHandle> getColumnHandles return ImmutableMap,
Map interface does not provide ordering guarantees.

So far it was not an issue. What are the benefits of this new method?
